### ISSUE TITLE
Onnx export fails on torch.distributions.Normal

### ISSUE THREAD

**Issue by dineNshine** [pytorch/pytorch#30517]
## 🐛 Bug

Onnx export fails on `torch.distributions.Normal.log_prob` calls because of the `broadcast_all` function called in `log_prob`.

## To Reproduce

```
import math
import torch
import torch.nn as nn
from torch.distributions.utils import broadcast_all


class Model(nn.Module):
  def __init__(self):
    super(Model, self).__init__()
    self.mean = nn.Parameter(torch.zeros((2, 10)), requires_grad=False)
    self.std = nn.Parameter(torch.ones((2, 10)), requires_grad=False)

  def forward(self, x):
    mean, std = broadcast_all(self.mean, self.std)
    # mean, std = self.mean, self.std
    return -((x - mean) ** 2) / (2 * std ** 2) - math.log(math.sqrt(2 * math.pi))


model = Model()
in_data = torch.ones((2, 10))
out_data = model(in_data)
torch.onnx.export(model, in_data, 'test.onnx', verbose=True)
```

Output:
```
/usr/local/lib/python3.5/dist-packages/torch/onnx/utils.py:617: UserWarning: ONNX export failed on ATen operator broadcast_tensors because torch.onnx.symbolic_opset9.broadcast_tensors does not exist
  .format(op_name, opset_version, op_name))
Traceback (most recent call last):
  File "onnx_export_test.py", line 24, in <module>
    torch.onnx.export(model, in_data, 'test.onnx', verbose=True)
  File "/usr/local/lib/python3.5/dist-packages/torch/onnx/__init__.py", line 143, in export
    strip_doc_string, dynamic_axes, keep_initializers_as_inputs)
  File "/usr/local/lib/python3.5/dist-packages/torch/onnx/utils.py", line 66, in export
    dynamic_axes=dynamic_axes, keep_initializers_as_inputs=keep_initializers_as_inputs)
  File "/usr/local/lib/python3.5/dist-packages/torch/onnx/utils.py", line 382, in _export
    fixed_batch_size=fixed_batch_size)
  File "/usr/local/lib/python3.5/dist-packages/torch/onnx/utils.py", line 262, in _model_to_graph
    fixed_batch_size=fixed_batch_size)
  File "/usr/local/lib/python3.5/dist-packages/torch/onnx/utils.py", line 132, in _optimize_graph
    graph = torch._C._jit_pass_onnx(graph, operator_export_type)
  File "/usr/local/lib/python3.5/dist-packages/torch/onnx/__init__.py", line 174, in _run_symbolic_function
    return utils._run_symbolic_function(*args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/torch/onnx/utils.py", line 618, in _run_symbolic_function
    op_fn = sym_registry.get_registered_op(op_name, '', opset_version)
  File "/usr/local/lib/python3.5/dist-packages/torch/onnx/symbolic_registry.py", line 91, in get_registered_op
    return _registry[(domain, version)][opname]
KeyError: 'broadcast_tensors'
```

Commenting the first line and uncommenting the second line in the `forward` function makes the module onnx exportable.

## Environment

 - PyTorch Version (e.g., 1.0): 1.3.1
 - OS (e.g., Linux): Ubuntu 16.04
 - How you installed PyTorch (`conda`, `pip`, source): pip3
 - Python version: 3.5.2
 - CUDA version: 10.2 (the 10.1 deb local and network installers install 10.2)
 - cuDNN version: 7.6
 - GPU models and configuration: 2 x GTX 1080 Ti

cc @ezyang @gchanan @zou3519 @jerryzh168 @vincentqb @fritzo @neerajprad @alicanb @vishwakftw @houseroad @spandantiwari @lara-hdr @BowenBao @neginraoof

**Comment by fritzo** [comment]
It looks like onnx needs to add support for the `torch.broadcast_tensors()` op.

**Comment by itssimon** [comment]
Still not supported :-(

`RuntimeError: Exporting the operator broadcast_tensors to ONNX opset version 11 is not supported. Please open a bug to request ONNX export support for the missing operator.`

Anything we can do to expedite the implementation of this operator?

**Comment by mxkrn** [comment]
Bump on this, interested in seeing support for this op.

`Exporting the operator broadcast_tensors to ONNX opset version 12 is not supported. Please open a bug to request ONNX export support for the missing operator.`

**Comment by lukasheinrich** [comment]
just to post that this is also important for us over at the pyhf project https://github.com/scikit-hep/pyhf

**Comment by itssimon** [comment]
Is anyone able to point out what needs to be changed/implemented and where (in ONNX or PyTorch) for this to be supported?

It appears that ONNX [generally supports broadcast](https://github.com/onnx/onnx/issues/3339#issuecomment-800791354), just not in this particular scenario.

**Comment by IcarusWizard** [comment]
Maybe not a neat solution, but we can overwrite the `Normal` class to use exportable operators.
```python
def exportable_broadcast(tensor1 : torch.Tensor, tensor2 : torch.Tensor):
    ''' broadcast tensors to the same shape using onnx exportable operators '''
    if len(tensor1.shape) < len(tensor2.shape):
        tensor2, tensor1 = exportable_broadcast(tensor2, tensor1)
    else:
        shape1 = tensor1.shape
        shape2 = tensor2.shape
        if len(shape1) == len(shape2):
            final_shape = [max(s1, s2) for s1, s2 in zip(shape1, shape2)]
            tensor1 = tensor1.expand(*final_shape)
            tensor2 = tensor2.expand(*final_shape)
        else:
            tensor2 = tensor2.expand(*shape1)
    return tensor1, tensor2

class ExportableNormal(Normal):
    def __init__(self, loc, scale, validate_args):
        self.loc, self.scale = exportable_broadcast(loc, scale)
        batch_shape = self.loc.size()
        super(Normal, self).__init__(batch_shape, validate_args=validate_args)
```

**Comment by garymm** [comment]
This seems fixed on PyTorch 1.10.0. The code from OP prints:
```
graph(%0 : Float(2, 10, strides=[10, 1], requires_grad=0, device=cpu),
      %mean : Float(2, 10, strides=[10, 1], requires_grad=0, device=cpu),
      %std : Float(2, 10, strides=[10, 1], requires_grad=0, device=cpu),
      %24 : Long(2, strides=[1], requires_grad=0, device=cpu),
      %25 : Float(requires_grad=0, device=cpu),
      %26 : Float(requires_grad=0, device=cpu)):
  %4 : Float(2, 10, device=cpu) = onnx::ConstantOfShape[value={0}](%24)
  %5 : Float(2, 10, strides=[10, 1], device=cpu) = onnx::Add(%4, %mean)
  %6 : Float(2, 10, strides=[10, 1], device=cpu) = onnx::Add(%5, %std)
  %7 : Long(2, strides=[1], device=cpu) = onnx::Shape(%6)
  %8 : Float(2, 10, device=cpu) = onnx::Expand(%mean, %7)
  %9 : Long(2, strides=[1], device=cpu) = onnx::Shape(%6)
  %10 : Float(2, 10, device=cpu) = onnx::Expand(%std, %9)
  %11 : Float(2, 10, strides=[10, 1], requires_grad=0, device=cpu) = onnx::Sub(%0, %8) # <ipython-input-16-c6474da21183>:16:0
  %14 : Float(2, 10, strides=[10, 1], requires_grad=0, device=cpu) = onnx::Pow(%11, %25) # /usr/local/lib/python3.7/dist-packages/torch/_tensor.py:30:0
  %15 : Float(2, 10, strides=[10, 1], requires_grad=0, device=cpu) = onnx::Neg(%14) # <ipython-input-16-c6474da21183>:16:0
  %18 : Float(2, 10, strides=[10, 1], requires_grad=0, device=cpu) = onnx::Pow(%10, %26) # /usr/local/lib/python3.7/dist-packages/torch/_tensor.py:30:0
  %19 : Float(requires_grad=0, device=cpu) = onnx::Constant[value={2}]()
  %20 : Float(2, 10, strides=[10, 1], requires_grad=0, device=cpu) = onnx::Mul(%18, %19)
  %21 : Float(2, 10, strides=[10, 1], requires_grad=0, device=cpu) = onnx::Div(%15, %20) # <ipython-input-16-c6474da21183>:16:0
  %22 : Float(requires_grad=0, device=cpu) = onnx::Constant[value={0.918939}]()
  %23 : Float(2, 10, strides=[10, 1], requires_grad=0, device=cpu) = onnx::Sub(%21, %22)
  return (%23)
```
